teps/0147-dynamic-matrix.md

```diff
@@ -0,0 +1,379 @@
+significance in how combinations are being generated. `name` is useful from the readability perspective with
+the explicit combinations when specified in the pipeline definition itself. With this proposal, `name` is not included.
+The reason behind not includeing `name` is, JSON specifications inherently does not support an optional fields.
+The users have to specify empty string at the minimum.
+
+### `matrix.strategy` initialized by the `pipelineRun`
+
+The `Pipeline` addressing this use case will be defined as shown below:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Pipeline
+metadata:
+  name: pipeline-to-build-images-from-a-single-repo
+spec:
+  params:
+    - name: matrix-strategy
+  workspaces:
+    - name: shared-workspace
+  tasks:
+    - ...
+    - name: kaniko-build
+      taskRef:
+        name: kaniko
+      workspaces:
+        - name: source
+          workspace: shared-workspace
+      matrix:
+        strategy: $(params.matrix-strategy)
+```
+
+A `PipelineRun` can be created to execute the above `pipeline` to build 1 image:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: PipelineRun
+metadata:
+  generateName: pipelineRun-
+spec:
+  pipelineRef:
+    name: pipeline-to-build-images-from-a-single-repo
+  params:
+    - name: matrix-strategy
+      value: "{\"include\":[{\"IMAGE\":\"image-1\",\"DOCKERFILE\":\"path/to/Dockerfile1\"}]}"
+```
+
+`PipelineRun` can be created to execute the above `pipeline` to build 3 images:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: PipelineRun
+metadata:
+  generateName: pipelineRun-
+spec:
+  pipelineRef:
+    name: pipeline-to-build-images-from-a-single-repo 
+  params:
+    - name: matrix-strategy
+      value: "{\"include\":[{\"IMAGE\":\"image-1\",\"DOCKERFILE\":\"path/to/Dockerfile1\"},{\"IMAGE\":\"image-2\",\"DOCKERFILE\":\"path/to/Dockerfile2\"},{\"IMAGE\":\"image-3\",\"DOCKERFILE\":\"path/to/Dockerfile3\"}]}"
+```
+
+### `matrix.strategy` specified by a task result
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Pipeline
+metadata:
+  name: pipeline-to-build-images-from-a-single-repo
+spec:
+  workspaces:
+    - name: shared-workspace
+  tasks:
+    - ...
+    - name: pre-requisite
+      taskSpec:
+        results:
+          - name: matrix-strategy
+        steps:
+          - name: read-strategy
+            image: alpine
+            script: |
+              # this strategy can be read from a JSON file in the repo
+              echo -n "{\"include\":[{\"IMAGE\":\"image-1\",\"DOCKERFILE\":\"path/to/Dockerfile1\"}]}" | tee $(results.matrix-strategy.path)
+    - name: kaniko-build
+      params:
+        - name: matrix-strategy
+          value: $(tasks.read-matrix-strategy.results.matrix-strategy)
+      taskRef:
+        name: kaniko
+      workspaces:
+        - name: source
+          workspace: shared-workspace
+      matrix:
+        strategy: $(params.matrix-strategy)
```

Comment:
It's not clear to me why the result is passed to a pt.param then passed to pt.matrix.strategy, I'd have expected the result to be consumed directly in pt.matrix.strategy just like they are in pt.when for when expressions

```suggestion
      taskRef:
        name: kaniko
      workspaces:
        - name: source
          workspace: shared-workspace
      matrix:
        strategy: $(tasks.read-matrix-strategy.results.matrix-strategy)
```

Reply:
yup, its up to the users how they would like to design their pipelines, both `params` and `results` are supported.